### PR TITLE
Use a transparent color when setting fill/stroke colors in a pattern context but with no colorspace

### LINKED ISSUE #18466: [Bug]: Invalid rendering because of no colorspace
### Attach (recommended) or Link to PDF file

[image2.pdf](https://github.com/user-attachments/files/16319966/image2.pdf)

### Web browser and its version

Firefox nightly

### Operating system and its version

Windows 11

### PDF.js version

last one

### Is the bug present in the latest PDF.js version?

Yes

### Is a browser extension

No

### Steps to reproduce the problem

Open the pdf

### What is the expected behavior?

image

### What went wrong?

image

### Link to a viewer

_No response_

### Additional context

The bug has been reported:
https://issues.apache.org/jira/browse/PDFBOX-5851

As mentioned in the bug report, `scn` and `SCN` are used without a previous call to `cs` so we fallback on `DeviceGray`:
https://github.com/mozilla/pdf.js/blob/ed83d7c5e16798a56c493d56aaa8200dd280bb17/src/core/evaluator.js#L4876-L4877

We should probably have a `null` rgb color or something like that to indicate to use a fully transparent one.

The rendering is correct in either Acrobat or Chrome.

### PR DESCRIPTION
**Pull request by calixteman**

### CODE REVIEW

File: src/core/evaluator.js [hunk 1]
```diff
@@ -4873,8 +4883,26 @@ class EvalState {
     this.ctm = new Float32Array(IDENTITY_MATRIX);
     this.font = null;
     this.textRenderingMode = TextRenderingMode.FILL;
-    this.fillColorSpace = ColorSpace.singletons.gray;
-    this.strokeColorSpace = ColorSpace.singletons.gray;
+    this._fillColorSpace = ColorSpace.singletons.gray;
+    this._fillColorSpacestrokeColorSpace = ColorSpace.singletons.gray;
```
Comment (Snuffleupagus) [inline src/core/evaluator.js:4887]: Could this typo may explain the test failures?
```suggestion
    this._strokeColorSpace = ColorSpace.singletons.gray;
```

File: src/core/evaluator.js [hunk 2]
```diff
@@ -2007,9 +2007,8 @@ class PartialEvaluator {
                   localColorSpaceCache,
                 })
                 .then(function (colorSpace) {
-                  if (colorSpace) {
-                    stateManager.state.fillColorSpace = colorSpace;
-                  }
+                  stateManager.state.fillColorSpace =
+                    colorSpace || ColorSpace.singletons.gray;
```
Comment (Snuffleupagus) [inline src/core/evaluator.js:2011]: If this is necessary, should we not make the same sort of change below as well if nothing else than for consistency?
https://github.com/mozilla/pdf.js/pull/18467/files#diff-cf40014d9c7d352f3ae212e7fffe5386decf96a10e53066a4d900c3eb8fc559eR2035-R2037

Comment (calixteman) [reply src/core/evaluator.js:2011]: I wondered the same and I don't know. I'm even not sure that fallbacking on gray is a good choice in general, at least it fixes the 2 test failures in the previous run.

Comment (Snuffleupagus) [reply src/core/evaluator.js:2011]: Well, as mentioned in https://github.com/mozilla/pdf.js/issues/18466#issuecomment-2241107239 it seems that `DeviceGray` should be the default value.

So, for consistency, we do need to update the `OPS.setStrokeColorSpace` case below as well.

**Comment by calixteman** [comment]
There is a regression in issue6707.pdf.
It's due to the fact we've a previous valid colorspace which is not replaced by the default gray one.
On master, I just applied this diff:
```diff
diff --git a/src/core/colorspace.js b/src/core/colorspace.js
index 74b43a1c4..ee2e8493c 100644
--- a/src/core/colorspace.js
+++ b/src/core/colorspace.js
@@ -396,7 +396,8 @@ class ColorSpace {
               }
             }
           }
-          throw new FormatError(`Unrecognized ColorSpace: ${cs.name}`);
+          return this.singletons.gray;
+          //throw new FormatError(`Unrecognized ColorSpace: ${cs.name}`);
       }
     }
     if (Array.isArray(cs)) {
```
and interestingly the above pdf shows stuff which was invisible before.
I know this pdf is a bit corrupted, but I do wonder if instead of throwing in `ColorSpace::_parse` when we've a wrong cs we could fallback on `DeviceGray`.
In the last months, we removed few `throw ...` to just fallback on something (or sometimes just to do nothing), maybe we could do the same here.
@Snuffleupagus wdyt ?

**Comment by Snuffleupagus** [comment]
> I know this pdf is a bit corrupted, but I do wonder if instead of throwing in `ColorSpace::_parse` when we've a wrong cs we could fallback on `DeviceGray`.

Without testing I really don't know what the general effect would be if we change all the relevant cases in that method to just "swallow" errors, since it might "fix" some things and break others.

Maybe we should just update this patch to address the referenced issue, without regressing anything, and then we could look into those ideas in a follow-up instead to reduce the number of "moving parts" in this PR?

File: src/core/evaluator.js (hunk 2 above)
Comment (Snuffleupagus) [inline src/core/evaluator.js:2011]: I believe that we could fix all regressions with the following changes.
```suggestion
                  if (colorSpace) {
                    stateManager.state.fillColorSpace = colorSpace;
                  } else if (!stateManager.state.patternFillColorSpace) {
                    stateManager.state.fillColorSpace = ColorSpace.singletons.gray;
                  }
```

File: src/core/evaluator.js [hunk 3]
```diff
@@ -2033,9 +2032,8 @@ class PartialEvaluator {
                   localColorSpaceCache,
                 })
                 .then(function (colorSpace) {
-                  if (colorSpace) {
-                    stateManager.state.strokeColorSpace = colorSpace;
-                  }
+                  stateManager.state.strokeColorSpace =
+                    colorSpace || ColorSpace.singletons.gray;
```
Comment (Snuffleupagus) [inline src/core/evaluator.js:2036]: Similar to the above.
```suggestion
                  if (colorSpace) {
                    stateManager.state.strokeColorSpace = colorSpace;
                  } else if (!stateManager.state.patternStrokeColorSpace) {
                    stateManager.state.strokeColorSpace = ColorSpace.singletons.gray;
                  }
```

**Review by Snuffleupagus — APPROVED** [review]
r=me, thank you!

**Pull request merged**